codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/awssdk/shapevisitor/ShapeVisitorHelper.java

```diff
@@ -0,0 +1,138 @@
+package software.amazon.polymorph.smithygo.awssdk.shapevisitor;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import software.amazon.polymorph.smithygo.codegen.GenerationContext;
+import software.amazon.polymorph.smithygo.codegen.GoWriter;
+import software.amazon.polymorph.smithygo.localservice.nameresolver.DafnyNameResolver;
+import software.amazon.polymorph.traits.ReferenceTrait;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.EnumTrait;
+
+public class ShapeVisitorHelper {
+
+  public static final Map<MemberShape, Boolean> toDafnyOptionalityMap =
+    new HashMap<>();
+  public static final Map<MemberShape, Boolean> toNativeOutputPointerMap =
+    new HashMap<>();
+  public static final Map<MemberShape, String> toNativeNextFuncInput =
+    new HashMap<>();
+
+  /**
+   * Generates functions Name for To Dafny and To Native conversion.
+   *
+   * @param memberShape       MemberShape to generate function name for.
+   * @param suffix            Suffix to add to the function. As of this writing, we only put FromDafny or ToNative suffix.
+   * @return the function Name
+   */
+  public static String funcNameGenerator(
+    final MemberShape memberShape,
+    final String suffix
+  ) {
+    return memberShape
+      .getId()
+      .toString()
+      .replaceAll("[.$#]", "_")
+      .concat("_")
+      .concat(suffix);
+  }
+
+  public static String toNativeShapeVisitorWriter(
+    final MemberShape memberShape,
+    final GenerationContext context,
+    final String dataSource,
+    final Boolean assertionRequired,
+    final GoWriter writer,
+    final boolean isOptional,
+    final boolean isPointable
+  ) {
+    final Shape targetShape = context
+      .model()
+      .expectShape(memberShape.getTarget());
+    String maybeAssertion = "";
+    if (assertionRequired && !targetShape.hasTrait(EnumTrait.class)) {
+      maybeAssertion =
+        ".(".concat(
+            DafnyNameResolver.getDafnyType(
+              targetShape,
+              context.symbolProvider().toSymbol(targetShape)
+            )
+          )
+          .concat(")");
+    }
+    String nextVisitorFunction;
+    String funcDataSource = "input";
+    if (!DafnyToAwsSdkShapeVisitor.visitorFuncMap.containsKey(memberShape)) {
+      DafnyToAwsSdkShapeVisitor.visitorFuncMap.put(memberShape, "");
+      DafnyToAwsSdkShapeVisitor.visitorFuncMap.put(
```

Comment:
Why are we updating the same key `memberShape` twice?

Reply:
Its because even before putting to the map, we are visiting shape. And in other shape that is visited we are going to do if check on `DafnyToAwsSdkShapeVisitor.visitorFuncMap.containsKey(memberShape)`, in case of recursive shape this statement will be false and the same shape will be again visited. So, to avoid that a empty string is send first and then the actual value